src/actions/pkg.cr

```diff
@@ -1,45 +1,142 @@
+        selected_pkg = DppmRestApi.prefix.new_pkg package_name, version
+        raise NoSuchPackage.new context, package_name unless selected_pkg.exists?
+        # Stop here ^^ unless the package whose name was the :id URL parameter
+        # was found and we can query it
+        if key = context.params.query["get"]?
+          data = begin
+            selected_pkg.get_config key
+          rescue error : ConfigKeyError
+            raise NotFound.new context, cause: error
+          end
+          # A key was specified by the &get query parameter
+          build_json context.response do |json|
+            json.field package_name do
+              json.object do
+                json.field key, value: data
+              end
+            end
+          end
+          next context
+        else
+          # No specific key was requested, respond with all of the config
+          # options for this package. Additionally, respond with all
+          # dependent packages if the "libraries" boolean query parameter
+          # is specified.
+          build_json context.response do |builder|
+            build_config_response selected_pkg, builder
+            if parse_boolean_param("libraries", from: context) &&
+               (libs = selected_pkg.libs)
+              builder.field "libraries" do
+                libs.each { |pkg| build_config_response pkg, builder }
+              end
+            end
+          end
+        end
+        next context
+      end
+      raise Unauthorized.new context
+    end
+    # Delete a given package
+    relative_delete "/:id/delete" do |context|
+      if Actions.has_access? context, Access::Delete
+        package_name = URI.unescape context.params.url["id"]
+        selected_pkg = DppmRestApi.prefix.new_pkg package_name,
+          context.params.query["version"]?
+        raise NoSuchPackage.new context, package_name if selected_pkg.nil?
+        selected_pkg.delete confirmation: false { }
+        build_json context.response do |json|
+          json.field "status", "successfully deleted '#{package_name}'"
+        end
+        next context
+      end
+      raise Unauthorized.new context
+    end
+
+    # Build a package, returning the ID of the built image, and perhaps a status
+    # message? We could also use server-side events or a websocket to provide the
+    # status of this action as it occurs over the API, rather than just returning
+    # a result on completion.
+    #
+    # This route takes the optional query parameters "version" and "tag".
+    relative_post "/:id/build" do |context|
+      if Actions.has_access? context, Access::Create
+        package_name = URI.unescape context.params.url["id"]
+        pfx = DppmRestApi.prefix
+        pfx.update
+        pkg = pfx.new_pkg package_name, version: context.params.query["version"]?
+        init_done = false
+        begin
+          pkg.build confirmation: false { init_done = true }
+        rescue e
```

Comment:
This still gives an *Internal Server Error* instead of *Bad Request* when given a package name that does not exist. In fact, I don't see any scenario in which an exception will be raised before [the block is yielded](https://github.com/DFabric/dppm/blob/fa4faae67e475f87c820020854ce71c6309f6074/src/prefix/pkg.cr#L143-L150).

Reply:
Good point, sorry